### PR TITLE
[SPARK-40245][SQL] Fix FileScan equality check when partition or data filter columns are not read

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
Unfortunately the fix in https://github.com/apache/spark/pull/31848 was not correct in all cases. When the partition or data filter contains a column that is not in `readSchema()` the filter nornalization in `FileScan.equals()` doesn't work.

### Why are the changes needed?
To fix `FileScan.equals()` to fix reuse issues.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new UT.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -4003,6 +4003,36 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+          .withColumn("p", $"id" % 2)
+          .write
+          .mode("overwrite")
+          .partitionBy("p")
+          .parquet(path.toString)
+        withTempView("t") {
+          spark.read.parquet(path.toString).createOrReplaceTempView("t")
+          val df = sql(
+            """
+              |SELECT t1.id, t2.id, t3.id
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4020]: do we return wrong result before?

Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4020]: No, just exchange reuse didn't happen.

**Comment by cloud-fan** [comment]
LGTM, please fix conflicts

**Comment by peter-toth** [comment]
> LGTM, please fix conflicts

Ok, merged latest `master` into this PR.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by peter-toth** [comment]
Thanks for the review!

**Pull request closed without merging**